[Site Isolation][iOS] Send touch events to isolated frames

charliewolfe opened the pull request:
#### 355bf0478f6ed4d5f7bbb0543140776b8af74dea
<pre>
[Site Isolation][iOS] Send touch events to isolated frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=264998">https://bugs.webkit.org/show_bug.cgi?id=264998</a>
<a href="https://rdar.apple.com/116837559">rdar://116837559</a>

Reviewed by Alex Christensen.

This patch makes changes to send touch events to isolated frames with corrected coordinates.

Similar to mouse and drag events, IPC will be sent between the UI process and web processes until the
target frame is reached. More details below.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::userInputEventDataForRemoteFrame):
(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):
(WebCore::EventHandler::handleTouchEvent):
(WebCore::EventHandler::dispatchSyntheticTouchEventIfEnabled):
(WebCore::EventHandler::mouseEventDataForRemoteFrame): Deleted.
* Source/WebCore/page/EventHandler.h:
(WebCore::EventHandler::touchEventTargetSubframe const):

I renamed `mouseEventDataForRemoteFrame()` to `userInputEventDataForRemoteFrame()` since this function
won’t only be used for mouse events.

* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::dispatchSimulatedTouchEvent):
(WebCore::EventHandler::touchEvent):

* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebTouchEvent::setPosition):

We need `setPosition()` to correct the coordinates of the touch event before forwarding it to the iframe
process.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::handlePreventableTouchEvent):
(WebKit::WebPageProxy::sendUnpreventableTouchEvent):
(WebKit::WebPageProxy::handleUnpreventableTouchEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

I added `sendPreventableTouchEvent()` and `sendUnpreventableTouchEvent()` so we can specify a frame to
send the touch event to. If the touch event needs to be forwarded to an iframe process the event
coordinates will be corrected and we’ll send another IPC call to the target frame.

* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::touchEvent):
(WebKit::EventDispatcher::touchEventWithoutCallback): Deleted.
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/EventDispatcher.messages.in:

We now need a completion handler even if the touch event is unpreventable, so I removed
`touchEventWithoutCallback()` and made both preventable and unpreventable touch events call directly
into `touchEvent()`. If the touch event is unpreventable the `handled` bool passed back through the
completion handler will be ignored. I also needed to store the completion handler for unpreventable
touch events in the touch event queue.

I added a frame identifier to be associated with each queued touch event. Also store these values in a
new struct.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::handleTouchEvent):
(WebKit::WebPage::dispatchTouchEvent):
(WebKit::WebPage::touchEvent):
(WebKit::WebPage::updatePotentialTapSecurityOrigin):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Change these functions to expect a frame identifier to be passed, and use the event handler associated
with that frame to handle touch events.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::dispatchAsynchronousTouchEvents):
(WebKit::WebPage::cancelAsynchronousTouchEvents):

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/site-isolation/touch-events/resources/post-message-coordinates.html: Added.
* LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates.html: Added.

This test verifies the isolated iframe receives the touch events with the expected coordinates. Also
make sure touch event listeners on the main frame are not called.

Canonical link: <a href="https://commits.webkit.org/270954@main">https://commits.webkit.org/270954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1761bb4489128e8557237286b5589106c62e75e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26693 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28905 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24407 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2703 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24316 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26955 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22915 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3624 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23907 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29390 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24355 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24306 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29931 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3702 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27830 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5150 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23643 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4171 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3491 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1142,7 +1142,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void resetPotentialTapSecurityOrigin();
     void handlePreventableTouchEvent(NativeWebTouchEvent&);
-    void handleUnpreventableTouchEvent(const NativeWebTouchEvent&);
+    void handleUnpreventableTouchEvent(NativeWebTouchEvent&);
```
Can we add const to updateTouchEventTracking instead of removing const here?

charliewolfe replied:
I don't think so? I needed to remove const here because we are passing the event to `sendUnpreventableTouchEvent` which may modify the event in the lambda. I guess I could create a mutable copy of the event so we keep const. I'll do that.

cdumez replied:
> I don't think so? I needed to remove const here because we are passing the event to sendUnpreventableTouchEvent which may modify the event in the lambda.

Note that the event in the lambda is a copy of the event you passed to the function, because you captured it by value. Therefore that inner event is not const as long as the lambda is marked as mutable.

charliewolfe replied:
> Note that the event in the lambda is a copy of the event you passed to the function, because you captured it by value. Therefore that inner event is not const as long as the lambda is marked as mutable.

I don't think that's right. `event` doesn't lose const after being passed into the lambda capture by value, I get a compile error when I try. If I pass it as `[event = event]` it does work.

achristensen07 commented on Source/WebKit/Shared/WebTouchEvent.h:
```diff
@@ -130,6 +130,7 @@ class WebTouchEvent : public WebEvent {
     const Vector<WebPlatformTouchPoint>& touchPoints() const { return m_touchPoints; }
 
     WebCore::IntPoint position() const { return m_position; }
+    void setPosition(WebCore::IntPoint position) { m_position = position; }
```
setTransformedPosition?

charliewolfe replied:
I named this `setPosition()` to match the function in `WebMouseEvent`. I can change both of these but maybe I shouldn't in this PR?

achristensen07 commented on LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates.html:
```diff
@@ -0,0 +1,28 @@
+addEventListener("message", (event) => {
+    events.push(event.data);
+    if (events.length == 3) {
+        if (events[0] == "touchstart 40,40" && events[1] == "touchmove 90,90" && events[2] == "touchend 90,90")
```
shouldBe('events[0]', 'touch start 40,40');
shouldBe(...)
...

That way if the test starts failing we'll see more useful information other than just that it failed.

cdumez commented on LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates-expected.txt:
```diff
@@ -0,0 +1,5 @@
+PASS iframe received touch events with correct coordinates.
```
This test is missing a `description()` call in JS

cdumez commented on LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates.html:
```diff
@@ -0,0 +1,28 @@
+<script>
+if (window.testRunner) {
+    jsTestIsAsync = true;
+    testRunner.dumpAsText();
```
Not needed since you included js-test.js

cdumez commented on LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates.html:
```diff
@@ -0,0 +1,28 @@
+<script src="/js-test-resources/js-test.js"></script>
+<script src="/js-test-resources/ui-helper.js"></script>
+<script>
+if (window.testRunner) {
```
This check is not needed.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3636,25 +3636,29 @@ void WebPage::cancelCurrentInteractionInformationRequest()
 {
-    RefPtr localMainFrame = dynamicDowncast<LocalFrame>(page->mainFrame());
-    if (!localMainFrame || !localMainFrame->view())
+    auto* frame = WebProcess::singleton().webFrame(frameID);
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3685,8 +3689,8 @@ void WebPage::updatePotentialTapSecurityOrigin(const WebTouchEvent& touchEvent,
-    while (auto subframe = touchEventTargetFrame->eventHandler().touchEventTargetSubframe())
-        touchEventTargetFrame = subframe;
+    while (RefPtr localSubframe = dynamicDowncast<LocalFrame>(touchEventTargetFrame->eventHandler().touchEventTargetSubframe()))
+        touchEventTargetFrame = localSubframe;
```
WTFMove(localSubframe)

charliewolfe committed:
[Site Isolation][iOS] Send touch events to isolated frames
https://bugs.webkit.org/show_bug.cgi?id=264998
rdar://116837559

Reviewed by Alex Christensen.

This patch makes changes to send touch events to isolated frames with corrected coordinates.

Similar to mouse and drag events, IPC will be sent between the UI process and web processes until the
target frame is reached. More details below.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::userInputEventDataForRemoteFrame):
(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):
(WebCore::EventHandler::handleTouchEvent):
(WebCore::EventHandler::dispatchSyntheticTouchEventIfEnabled):
(WebCore::EventHandler::mouseEventDataForRemoteFrame): Deleted.
* Source/WebCore/page/EventHandler.h:
(WebCore::EventHandler::touchEventTargetSubframe const):

I renamed `mouseEventDataForRemoteFrame()` to `userInputEventDataForRemoteFrame()` since this function
won’t only be used for mouse events.

* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::dispatchSimulatedTouchEvent):
(WebCore::EventHandler::touchEvent):

* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebTouchEvent::setPosition):

We need `setPosition()` to correct the coordinates of the touch event before forwarding it to the iframe
process.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::handlePreventableTouchEvent):
(WebKit::WebPageProxy::sendUnpreventableTouchEvent):
(WebKit::WebPageProxy::handleUnpreventableTouchEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

I added `sendPreventableTouchEvent()` and `sendUnpreventableTouchEvent()` so we can specify a frame to
send the touch event to. If the touch event needs to be forwarded to an iframe process the event
coordinates will be corrected and we’ll send another IPC call to the target frame.

* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::touchEvent):
(WebKit::EventDispatcher::touchEventWithoutCallback): Deleted.
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/EventDispatcher.messages.in:

We now need a completion handler even if the touch event is unpreventable, so I removed
`touchEventWithoutCallback()` and made both preventable and unpreventable touch events call directly
into `touchEvent()`. If the touch event is unpreventable the `handled` bool passed back through the
completion handler will be ignored. I also needed to store the completion handler for unpreventable
touch events in the touch event queue.

I added a frame identifier to be associated with each queued touch event. Also store these values in a
new struct.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::handleTouchEvent):
(WebKit::WebPage::dispatchTouchEvent):
(WebKit::WebPage::touchEvent):
(WebKit::WebPage::updatePotentialTapSecurityOrigin):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Change these functions to expect a frame identifier to be passed, and use the event handler associated
with that frame to handle touch events.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::dispatchAsynchronousTouchEvents):
(WebKit::WebPage::cancelAsynchronousTouchEvents):

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/site-isolation/touch-events/resources/post-message-coordinates.html: Added.
* LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/touch-events/touch-event-coordinates.html: Added.

This test verifies the isolated iframe receives the touch events with the expected coordinates. Also
make sure touch event listeners on the main frame are not called.

Canonical link: https://commits.webkit.org/270954@main
